06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+   - Wait for nodes to have `SERVING` within the `postOperationTimeoutMs`. 
+   - If the timeout is reached for a node and its `numRetries` is greater than or equal to `maxRetries`, throw `UnrestartableNodesException` with the log recovery progress (number of remaining logs and segments). Otherwise increment node's `numRetries` and repeat from step 2.
+
+6. Restart nodes in `RESTART_FIRST` category:
+   - if one or more nodes have `NOT_RUNNING` state, we first need to check 2 special conditions:
+      - If all of the nodes are mixed and are in `NOT_RUNNING` state, restart them in parallel to give the best chance of forming the quorum.
+      > This is to address the issue described in https://github.com/strimzi/strimzi-kafka-operator/issues/9426.
+
+      - If a node is in `NOT_RUNNING` state, the restart it only if it has `POD_HAS_OLD_REVISION` or `POD_UNRESPONSIVE` reason. This is because, if the node is not running at all, then restarting it likely won't make any difference unless the node is out of date.
+      > For example, if a pod is in pending state due to misconfigured affinity rule, there is no point restarting this pod again or restarting other pods, because that would leave them in pending state as well. If the user then fixed the misconfigured affinity rule, then we should detect that the pod has an old revision, therefore should restart it so that pod is scheduled correctly and runs.
+
+      - At this point either we started nodes or decided not to because nodes did not have `POD_HAS_OLD_REVISION` reason. Regardless, wait for nodes to have `SERVING` within `postOperationTimeoutMs`. If the timeout is reached and the node's `numRetries` is greater than or equal to `maxRetries`, throw `TimeoutException`. Otherwise increment node's `numRetries` and repeat from step 2. 
+
+   
+   - Otherwise the nodes will be attempted to restart one by one in the following order:
+      - Pure controller nodes
+      - Mixed nodes
+      - Broker only nodes
+      
+   - Wait for the restarted node to have `SERVING` within `postOperationTimeoutMs`. If the timeout is reached and the node's `numRetries` is greater than or equal to `maxRetries`, throw `TimeoutException`. Otherwise increment node's `numRetries` and repeat from step 2. 
+
+7. Further refine the broker nodes in `MAYBE_RECONFIGURE` group:
+   - Describe Kafka configurations for each node via Admin API and compare them against the desired configurations. This is essentially the same mechanism we use today for the current KafkaRoller.
+   - If a node has configuration changes and they can be dynamically updated, add the node into another group called `RECONFIGURE`.
+   - If a node has configuration changes but they cannot be dynamically updated, add the node into the `RESTART` group.
+   - If a node has no configuration changes, put the node into the `NOP` group.
+
+8. Reconfigure each node in `RECONFIGURE` group:
```

Comment:
We sort nodes that are in MAYBE_RECONFIGURE into RECONFIGURE or RESTART groups. The reason for doing the RECONFIGURE first, to see if we can successfully configure the node dynamically. If configuration update fails, then it will be put into `RESTART` group.

Reply:
Fair enough, but:
* The names still differ between here and the list of groups on lines 129-134
* A nit, but it might be easier to read if the order of the groups on lines 129-134 is the same as in which you process it